materialize-sql/type_mapping.go

```diff
@@ -79,34 +85,90 @@ func BuildProjections(spec *pf.MaterializationSpec_Binding) (keys, values []Proj
 func (p *Projection) AsFlatType() (_ FlatType, mustExist bool) {
 	mustExist = p.Inference.Exists == pf.Inference_MUST
 
-	var types []FlatType
-	for _, ty := range p.Inference.Types {
+	types, hasNull := jsonTypesToFlatTypes(effectiveJsonTypes(&p.Projection))
+	if hasNull {
+		mustExist = false
+	}
+
+	switch len(types) {
+	case 0:
+		return NEVER, false
+	case 1:
+		return types[0], mustExist
+	default:
+		return MULTIPLE, mustExist
+	}
+}
+
+func jsonTypesToFlatTypes(jsonTypes []string) ([]FlatType, bool) {
+	var flatTypes []FlatType
+	var hasNullType bool
+	for _, ty := range jsonTypes {
 		switch ty {
 		case "string":
-			types = append(types, STRING)
+			flatTypes = append(flatTypes, STRING)
 		case "integer":
-			types = append(types, INTEGER)
+			flatTypes = append(flatTypes, INTEGER)
 		case "number":
-			types = append(types, NUMBER)
+			flatTypes = append(flatTypes, NUMBER)
 		case "boolean":
-			types = append(types, BOOLEAN)
+			flatTypes = append(flatTypes, BOOLEAN)
 		case "object":
-			types = append(types, OBJECT)
+			flatTypes = append(flatTypes, OBJECT)
 		case "array":
-			types = append(types, ARRAY)
+			flatTypes = append(flatTypes, ARRAY)
 		case "null":
-			mustExist = false
+			hasNullType = true
 		}
 	}
 
-	switch len(types) {
-	case 0:
-		return NEVER, false
-	case 1:
-		return types[0], mustExist
-	default:
-		return MULTIPLE, mustExist
+	return flatTypes, hasNullType
+}
+
+// effectiveJsonTypes potentially "condenses" the list of JSON types from a projection's Inference
+// into fewer types. Currently this supports cases where the projected field has exactly two types
+// with one of them being a string. The string must have a "format" matching the JSON type of the
+// non-string type, and the format must be one of the allowed formats that Flow handles. Fields for
+// collection keys are not condensed.
+func effectiveJsonTypes(projection *pf.Projection) []string {
+	if projection.IsPrimaryKey || projection.Inference.String_ == nil {
+		return projection.Inference.Types
+	}
+
+	format := projection.Inference.String_.Format
+	if !SliceContains(format, effectiveTypeFormats) {
+		return projection.Inference.Types
+	}
+
+	applicableType, ok := singleNonStringField(projection.Inference.Types)
```

Comment:
the combination of logic here and in `singleNonStringField`, plus the separate separate `effectiveTypeFormats` const, is hard to follow as a reader and is a bit smelly. It feels like too much and unnecessary indirection.

Also, this appears to allow for some unintended cases that probably _shouldn't_ be mapped:
 * `{type: [string, number], format: integer}`
 * `{type: [string, integer], format: number}`

What about having this routine be very explicit, such as:

```
if !projection.IsPrimaryKey && projection.Inference.String_ != nil {
  if reflect.DeepEquals(projection.Inference.Types, []string{"integer", "string") && format == "integer" {
    return []string{"integer"}
  }
  if reflect.DeepEquals(projection.Inference.Types, []string{"number", "string") && format == "number" {
    return []string{"number"}
  }
}
```
?

Note that types are [always in fixed, sorted order](https://github.com/estuary/flow/blob/master/crates/json/src/schema/types.rs#L62-L71) so you can rely on that for slice comparison if you like.

To be really pedantic, we should probably also handle "fractional", which is a type we introduced that is a floating-point and _not_ an integer. It actually can come up in inference, through constructions like: 

```
type: number
not: {type: integer}
```

We do allow it to be specified directly as well, as a JSON schema extension.

Reply:
Seems reasonable to simplify. I had originally done it this way with the thought of making it really easy to extend to other formats. Being more explicit does simplify things a lot, and it still wouldn't be hard to handle more formats if we ever actually need to, so I think this is a good idea. I made the change like you suggested & everything continues to work well.

I'll work on getting the `fractional` type to work when specified directly.